apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/RetryOnRdsFailover.java

```diff
@@ -0,0 +1,31 @@
+package gov.cms.bfd.server.war.commons;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import org.springframework.retry.annotation.Backoff;
+import org.springframework.retry.annotation.Retryable;
+import software.amazon.jdbc.plugin.failover.FailoverSQLException;
+
+/**
+ * Annotation encapsulating common parameters for the {@link Retryable} spring-retry annotation for
+ * use with {@link ca.uhn.fhir.rest.annotation.Search} and {@link ca.uhn.fhir.rest.annotation.Read}
+ * annotated resource provider methods that need to retry on {@link FailoverSQLException}s.
+ *
+ * <p>Some providers wrap checked {@link Exception}s thrown by asynchronous tasks in {@link
+ * RuntimeException}s, thus using a simple "retryFor" expression matching on {@link
+ * FailoverSQLException} is not possible for all provider operations. Instead, a SpEL expression is
+ * used to get the root cause of the thrown {@link Exception} and see if it is an instance of {@link
+ * FailoverSQLException} (note that the root cause of a {@link FailoverSQLException} is itself).
+ */
+@Target({ElementType.METHOD, ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Retryable(
+    exceptionExpression =
+        "T(org.apache.commons.lang3.exception.ExceptionUtils).getRootCause(#root) instanceof T(software.amazon.jdbc.plugin.failover.FailoverSQLException)",
```

Comment:
This `exceptionExpression` uses [SpEL](https://docs.spring.io/spring-framework/reference/core/expressions.html) (Spring Expression Language). Essentially, this string will be evaluated as an expression at runtime for each `Exception` that `@RetryOnRdsFailover`-decorated methods throw. `#root`, in this context, is the `Exception` that was thrown, so this expression is essentially equivalent to:

```java
// Assume #root is ex
return ExceptionUtils.getRootCause(ex) instanceof FailoverSQLException
```

We must specify the fully-qualified path to each type/static class (using the `T(...)` syntax) because we cannot guarantee that these types are imported when this expression is evaluated. I do realize that this is not entirely ideal, and I could extract this expression out into a separate `SpringRetryUtils` static utility class or something similar if we're uncomfortable with this line.

Reply:
I'm assuming `@Retryable(retryFor = FailoverSqlException.class)` doesn't work because FailoverSqlException is the inner exception rather than the outer one?

Reply:
Would be nice if you could have it call a method instead of this stringly-typed business, but if this is the only choice, I think it's fine to keep it inline.

Reply:
> I'm assuming `@Retryable(retryFor = FailoverSqlException.class)` doesn't work because FailoverSqlException is the inner exception rather than the outer one?

Yep. There's some downstream methods called by the providers that throw, essentially: `new RuntimeException(new JDBCException(new FailoverSuccessSQLException)))`. Even if they don't wrap it in an outer `RuntimeException`, the resulting exception will almost always be a `JDBCException` wrapping a `FailoverSuccessSQLException`. So we can't rely on `retryFor` here because it only looks at the type of the outermost `Exception` for matching.

> Would be nice if you could have it call a method instead of this stringly-typed business, but if this is the only choice, I think it's fine to keep it inline.

I thought about it some more and I think I'm leaning more towards a `SpringRetryUtils` static class with a static method called something like `shouldRetryFailover(...)` that implements the same logic as this SpEL expression. We'd still need to use the canonical class name (i.e. `gov.cms.bfd.server.war.commons.SpringRetryUtils`) in the SpEL that calls it, though, something like:

```
exceptionExpression = "T(gov.cms.bfd.server.war.commons.SpringRetryUtils).shouldRetryFailover(#root)"
```

_But_, the `shouldRetryFailover` method could be unit tested and we could be more confident that the package path of `SpringRetryUtils` won't change under us without warning.

I'm going to refactor this following the above.

Reply:
Nice, I think that's a worthy improvement

Reply:
This has been done in [4e291f3](https://github.com/CMSgov/beneficiary-fhir-data/pull/2402/commits/4e291f3b1f6b2d3c9d4a18011209cb8132266e86)